### PR TITLE
Controlled operations rework Part 1

### PR DESCRIPTION
**Pull request by astralcai**
**Context:**
All controlled operations should inherit from the general Controlled class, and the decomposition of controlled operations is not consistent for custom and non-custom controlled operations. This is a continuation of https://github.com/PennyLaneAI/pennylane/pull/5069

This is the first PR out of two for this rework. The second PR will focus on making sure that all custom controlled operations inherit from Controlled for more consistent inheritance structure.

**Description of the Change:**
- Make `MultiControlledX` inherit from ControlledOp.
- `qml.ctrl` called on operators with custom controlled versions will return instances of the custom class.
- Special handling of `PauliX` based controlled operations (`PauliX`, `CNOT`, `Toffoli`, `MultiControlledX`)
  - Calling `qml.ctrl` on one of these operators will always resolve to the best option in `CNOT`, `Toffoli`, or `MultiControlledX` depending on the number of control wires and control values.
- `qml.ctrl` will flatten nested controlled operators to a single multi-controlled operation.
- Controlled operators with a custom controlled version decomposes like how their controlled counterpart decomposes, as opposed to decomposing into their controlled version.
  - Special handling of `PauliX` based controlled operations: e.g., `Controlled(CNOT([0, 1]), [2, 3])` will have the same decomposition behaviour as a `MultiControlledX([2, 3, 0, 1])`

**Benefits:**
Cleaner code and more consistent behaviour

**Possible Drawbacks:**
Change of decomposition behaviour may cause issues.
~For `MultiControlledX`, the `wires` attribute now refers to all wires, as in `control_wires + target_wire + work_wires`, to access only the `control_wires + target_wires`, use the `active_wires` attribute.~

**Related GitHub Issues:**
https://github.com/PennyLaneAI/pennylane/pull/5069
https://github.com/PennyLaneAI/pennylane/issues/1447

**Related Shortcut Stories**
[sc-55949]
[sc-55131]
[sc-55358]

### CODE REVIEW

File: pennylane/drawer/tape_mpl.py [hunk 1]
```diff
@@ -107,8 +107,7 @@ def _(op: ops.Toffoli, drawer, layer, _):
 
 @_add_operation_to_drawer.register
 def _(op: ops.MultiControlledX, drawer, layer, _):
-    control_values = [(i == "1") for i in op.hyperparameters["control_values"]]
```
[the remark this answers is not in the document]
Comment (astralcai) [reply pennylane/drawer/tape_mpl.py:110]: Yeah things like this make up an entire day of pain lol

File: tests/transforms/test_qmc_transform.py [hunk 2]
```diff
@@ -260,7 +260,7 @@ def fn():
 
         dev = qml.device("default.qubit", wires=wires + estimation_wires)
 
-        @qml.qnode(dev)
+        @qml.qnode(dev, expansion_strategy="device")
```
Comment (albi3ro) [inline tests/transforms/test_qmc_transform.py:263]:
```suggestion
        @qml.qnode(dev)
```

File: tests/ops/op_math/test_controlled_ops.py [hunk 3]
```diff
@@ -695,215 +693,6 @@ def test_eigvals_tf(self, tol, op, params, expected_eigvals):
         )
 
 
-def _dot_broadcasted(a, b):
```
Comment (albi3ro) [inline tests/ops/op_math/test_controlled_ops.py:698]: Mind reminding me why these tests are getting removed?

Comment (astralcai) [reply tests/ops/op_math/test_controlled_ops.py:698]: They are all moved to `test_controlled.py`, as parameters in the `TestDecomposition` test class

File: tests/ops/op_math/test_controlled.py [hunk 4]
```diff
@@ -1608,146 +1913,288 @@ def test_qubit_unitary(M):
...
+        assert tape.expand(depth=2).circuit == [
+            qml.ControlledPhaseShift(np.pi / 4, wires=[3, 7]),
+            qml.Toffoli(wires=[3, 7, 0]),
+            qml.ControlledPhaseShift(-np.pi / 4, wires=[3, 0]),
+            qml.Toffoli(wires=[3, 7, 0]),
+            qml.ControlledPhaseShift(np.pi / 4, wires=[3, 0]),
+        ]
 
+    def test_adjoint_of_ctrl(self):
+        """Tests that adjoint(ctrl(fn)) and ctrl(adjoint(fn)) are equivalent"""
```
Comment (dime10) [inline tests/ops/op_math/test_controlled.py:2024]: Here we test their equivalence only in conjunction with `qml.simplify`, would it be reasonable to make them equivalent without a call to simplify?

File: tests/ops/op_math/test_controlled.py [hunk 5]
```diff
@@ -781,63 +796,228 @@ def test_sparse_matrix_format(self):
...
+    @pytest.mark.parametrize(
+        "target, decomp",
+        [
+            (
+                OpWithDecomposition(0.123, wires=[0, 1]),
+                [
+                    qml.CH(wires=[2, 0]),
+                    Controlled(qml.S(wires=1), control_wires=2),
+                    qml.CRX(0.123, wires=[2, 0]),
+                ],
```
Comment (dime10) [inline tests/ops/op_math/test_controlled.py:991]: Nice, this is equivalent to the test we talked about in https://github.com/PennyLaneAI/pennylane/pull/5113#discussion_r1468208850 right?

Comment (astralcai) [reply tests/ops/op_math/test_controlled.py:991]: Yes.

File: tests/ops/op_math/test_controlled.py (hunk 4 above)
Comment (astralcai) [reply tests/ops/op_math/test_controlled.py:2024]: They are actually not equivalent without the call to `qml.simplify`. I mean they are equivalent as in they have the same effect, but they wouldn't be the exact same two sequences of operations. This might be something worth investigating but it would be outside the scope of this PR. This particular test case is not new, it was copied over from another place.

File: pennylane/devices/qubit/apply_operation.py
Comment (albi3ro) [inline pennylane/devices/qubit/apply_operation.py:1]: We can revert the changes in this file now.

Comment (astralcai) [reply pennylane/devices/qubit/apply_operation.py:1]: Done

File: doc/releases/changelog-dev.md [hunk 6]
```diff
@@ -150,6 +150,9 @@
   and the codecov check itself would never execute.
   [(#5101)](https://github.com/PennyLaneAI/pennylane/pull/5101)
 
+* `qml.ctrl` called on operators with custom controlled versions will return instances of the custom class, and it will also flatten nested controlled operators to a single multi-controlled operation. For `PauliX`, `CNOT`, `Toffoli`, and `MultiControlledX`, calling `qml.ctrl` will always resolve to the best option in `CNOT`, `Toffoli`, or `MultiControlledX` depending on the number of control wires and control values.
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:153]:
```suggestion
* `qml.ctrl` called on operators with custom controlled versions will return instances
  of the custom class, and it will also flatten nested controlled operators to a single
  multi-controlled operation. For `PauliX`, `CNOT`, `Toffoli`, and `MultiControlledX`,
  calling `qml.ctrl` will always resolve to the best option in `CNOT`, `Toffoli`, or
  `MultiControlledX` depending on the number of control wires and control values.
```

File: doc/releases/changelog-dev.md [hunk 7]
```diff
@@ -185,7 +188,11 @@
   (with potentially negative eigenvalues) has been implemented.
   [(#5048)](https://github.com/PennyLaneAI/pennylane/pull/5048)
 
-* The decomposition of an operator created with calling `qml.ctrl` on a parametric operator (specifically `RX`, `RY`, `RZ`, `Rot`, `PhaseShift`) with a single control wire will now be the full decomposition instead of a single controlled gate. For example:
+* Controlled operators with a custom controlled version decomposes like how their controlled counterpart decomposes, as opposed to decomposing into their controlled version.
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:191]:
```suggestion
* Controlled operators with a custom controlled version decomposes like how their
  controlled counterpart decomposes, as opposed to decomposing into their controlled version.   
```

File: tests/ops/op_math/test_controlled.py (hunk 4 above)
Comment (dime10) [reply tests/ops/op_math/test_controlled.py:2024]: Yep agreed, it probably is it out of scope for this PR. I just brought it up here when I saw this test because it was something I also suggested to you in a our private conversation, but that could be of interest to other people.

Specifically, arbitrary chains of `C(A(C(C(C(A(...))))))` could be simplified automatically upon construction if we defined a **canonical ordering** for `Adjoint` and `Controlled`. Any such chain could then be simplified to at-most 1 adjoint and 1 control operator. For the given example, if the canonical order is for adjoint to be inside, we could automatically have the following steps take place:
```
-> C(A(C(A(...)))) (after the existing simplification of nested control)
-> C(C(A(A(...)))) 
-> C(C(A(...))) (if adjoint simplification is implemented?)
-> C(A(...))
```
Whereas with this PR we are stuck at the first step.

**Review by albi3ro — APPROVED** [review]
Looks great 🎉  Thanks for all your hard work :)

File: pennylane/ops/op_math/controlled_ops.py [hunk 8]
```diff
@@ -339,6 +344,242 @@ def compute_matrix():  # pylint: disable=arguments-differ
...
+    @staticmethod
+    def compute_decomposition(wires):  # pylint: disable=arguments-differ
+        return [qml.ControlledPhaseShift(np.pi, wires=wires)]
+
+    def decomposition(self):
+        return self.compute_decomposition(self.wires)
+
+
+def _check_and_convert_control_values(control_values, control_wires):
+    # TODO: maybe at some point actually deprecate string representations of control values?
```
Comment (timmysilv) [inline pennylane/ops/op_math/controlled_ops.py:356]: [scheduled](https://app.shortcut.com/xanaduai/story/49875/deprecate-passing-a-string-for-control-values-to-multicontrolledx) already for 0.36, so we can remove this TODO

File: tests/ops/op_math/test_controlled.py (hunk 4 above)
Comment (astralcai) [reply tests/ops/op_math/test_controlled.py:2024]: This looks nice, but at the same time if some users/developers still use the `Adjoint` constructor, then we won't be able to assume this canonical ordering everywhere unless we enforce this ordering in the `Adjoint` constructor instead.

File: pennylane/ops/op_math/controlled_decompositions.py
Comment (timmysilv) [inline pennylane/ops/op_math/controlled_decompositions.py:1]: did you change any of the decomposition helper functions in this file? it's hard to tell because of the move

**Review by timmysilv — COMMENTED** [review]
This looks great so far! most of my questions are style-related, but the bulk of this code looks super nice. a lot of work in this I see 🏆 

It's a little hard to see what's actually changed in MultiControlledX because of the file move, so I've generated the (reverse of the) diff of the class and attached it here:
<details>
<summary>The MCX diff, reversed</summary>

```diff
-class MultiControlledX(ControlledOp):
+class MultiControlledX(Operation):
     r"""MultiControlledX(control_wires, wires, control_values)
     Apply a Pauli X gate controlled on an arbitrary computational basis state.
 
@@ -385,8 +385,8 @@ class MultiControlledX(ControlledOp):
             Now users should use "wires" to indicate both the control wires and the target wire.
         wires (Union[Wires, Sequence[int], or int]): control wire(s) followed by a single target wire where
             the operation acts on
-        control_values (Union[bool, list[bool], int, list[int]]): The value(s) the control wire(s)
-                should take. Integers other than 0 or 1 will be treated as ``int(bool(x))``.
+        control_values (str): a string of bits representing the state of the control
+            wires to control on (default is the all 1s state)
         work_wires (Union[Wires, Sequence[int], or int]): optional work wires used to decompose
             the operation into a series of Toffoli gates
 
@@ -417,66 +417,80 @@ class MultiControlledX(ControlledOp):
     """
 
     is_self_inverse = True
-    """bool: Whether or not the operator is self-inverse."""
 
     num_wires = AnyWires
-    """int: Number of wires the operation acts on."""
 
     num_params = 0
     """int: Number of trainable parameters that the operator depends on."""
 
-    ndim_params = ()
-    """tuple[int]: Number of dimensions per trainable parameter that the operator depends on."""
-
-    name = "MultiControlledX"
+    grad_method = None
 
     def _flatten(self):
-        return (), (self.active_wires, tuple(self.control_values), self.work_wires)
+        hyperparameters = (
+            ("wires", self.wires),
+            ("control_values", self.hyperparameters["control_values"]),
+            ("work_wires", self.hyperparameters["work_wires"]),
+        )
+        return tuple(), hyperparameters
 
     @classmethod
     def _unflatten(cls, _, metadata):
-        return cls(wires=metadata[0], control_values=metadata[1], work_wires=metadata[2])
+        return cls(**dict(metadata))
 
     # pylint: disable=too-many-arguments
     def __init__(self, control_wires=None, wires=None, control_values=None, work_wires=None):
         if wires is None:
             raise ValueError("Must specify the wires where the operation acts on")
-        wires = wires if isinstance(wires, Wires) else Wires(wires)
-        if control_wires is not None:
+        if control_wires is None:
+            if len(wires) > 1:
+                control_wires = Wires(wires[:-1])
+                wires = Wires(wires[-1])
+            else:
+                raise ValueError(
+                    "MultiControlledX: wrong number of wires. "
+                    f"{len(wires)} wire(s) given. Need at least 2."
+                )
+        else:
+            wires = Wires(wires)
+            control_wires = Wires(control_wires)
+
             warnings.warn(
-                "The control_wires keyword will be removed soon. Use wires = (control_wires, "
-                "target_wire) instead. See the documentation for more information.",
-                UserWarning,
+                "The control_wires keyword will be removed soon. "
+                "Use wires = (control_wires, target_wire) instead. "
+                "See the documentation for more information.",
+                category=UserWarning,
             )
+
             if len(wires) != 1:
                 raise ValueError("MultiControlledX accepts a single target wire.")
-        else:
-            if len(wires) < 2:
-                raise ValueError(
-                    f"MultiControlledX: wrong number of wires. {len(wires)} wire(s) given. Need at least 2."
-                )
-            control_wires = wires[:-1]
-            wires = wires[-1:]
 
-        control_values = _check_and_convert_control_values(control_values, control_wires)
+        work_wires = Wires([]) if work_wires is None else Wires(work_wires)
+        total_wires = control_wires + wires
 
-        super().__init__(
-            qml.PauliX(wires=wires),
-            control_wires=control_wires,
-            control_values=control_values,
-            work_wires=work_wires,
-        )
+        if Wires.shared_wires([total_wires, work_wires]):
+            raise ValueError("The work wires must be different from the control and target wires")
+
+        if not control_values:
+            control_values = "1" * len(control_wires)
+
+        self.hyperparameters["control_wires"] = control_wires
+        self.hyperparameters["work_wires"] = work_wires
+        self.hyperparameters["control_values"] = control_values
+        self.total_wires = total_wires
+
+        super().__init__(wires=self.total_wires)
 
     def __repr__(self):
-        return f"MultiControlledX(wires={self.active_wires.tolist()}, control_values={self.control_values})"
+        return f'MultiControlledX(wires={list(self.total_wires._labels)}, control_values="{self.hyperparameters["control_values"]}")'
 
-    @property
-    def wires(self):
-        return self.active_wires
+    def label(self, decimals=None, base_label=None, cache=None):
+        return base_label or "X"
 
-    # pylint: disable=unused-argument, arguments-differ
+    # pylint: disable=unused-argument
     @staticmethod
-    def compute_matrix(control_wires, control_values=None, **kwargs):
+    def compute_matrix(
+        control_wires, control_values=None, **kwargs
+    ):  # pylint: disable=arguments-differ
         r"""Representation of the operator as a canonical matrix in the computational basis (static method).
 
         The canonical matrix is the textbook matrix representation that does not consider wires.
@@ -486,40 +500,57 @@ class MultiControlledX(ControlledOp):
 
         Args:
             control_wires (Any or Iterable[Any]): wires to place controls on
-            control_values (Union[bool, list[bool], int, list[int]]): The value(s) the control wire(s)
-                should take. Integers other than 0 or 1 will be treated as ``int(bool(x))``.
-
+            control_values (str): string of bits determining the controls
         Returns:
-            tensor_like: matrix representation
+           tensor_like: matrix representation
 
         **Example**
 
-        >>> print(qml.MultiControlledX.compute_matrix([0], 1))
+        >>> print(qml.MultiControlledX.compute_matrix([0], '1'))
         [[1. 0. 0. 0.]
          [0. 1. 0. 0.]
          [0. 0. 0. 1.]
          [0. 0. 1. 0.]]
-        >>> print(qml.MultiControlledX.compute_matrix([1], 0))
+        >>> print(qml.MultiControlledX.compute_matrix([1], '0'))
         [[0. 1. 0. 0.]
          [1. 0. 0. 0.]
          [0. 0. 1. 0.]
          [0. 0. 0. 1.]]
 
         """
+        if control_values is None:
+            control_values = "1" * len(control_wires)
+
+        if isinstance(control_values, str):
+            if len(control_values) != len(control_wires):
+                raise ValueError("Length of control bit string must equal number of control wires.")
+
+            # Make sure all values are either 0 or 1
+            if not set(control_values).issubset({"1", "0"}):
+                raise ValueError("String of control values can contain only '0' or '1'.")
 
-        control_values = _check_and_convert_control_values(control_values, control_wires)
-        padding_left = sum(2**i * int(val) for i, val in enumerate(reversed(control_values))) * 2
+            control_int = int(control_values, 2)
+        else:
+            raise ValueError("Control values must be passed as a string.")
+
+        padding_left = control_int * 2
         padding_right = 2 ** (len(control_wires) + 1) - 2 - padding_left
-        return block_diag(np.eye(padding_left), qml.PauliX.compute_matrix(), np.eye(padding_right))
+        cx = block_diag(np.eye(padding_left), PauliX.compute_matrix(), np.eye(padding_right))
+        return cx
 
-    def matrix(self, wire_order=None):
-        canonical_matrix = self.compute_matrix(self.control_wires, self.control_values)
-        wire_order = wire_order or self.wires
-        return qml.math.expand_matrix(
-            canonical_matrix, wires=self.active_wires, wire_order=wire_order
+    @property
+    def control_wires(self):
+        return self.wires[:~0]
+
+    def adjoint(self):
+        return MultiControlledX(
+            wires=self.wires,
+            control_values=self.hyperparameters["control_values"],
         )
 
-    # pylint: disable=unused-argument, arguments-differ
+    def pow(self, z):
+        return super().pow(z % 2)
+
     @staticmethod
     def compute_decomposition(wires=None, work_wires=None, control_values=None, **kwargs):
         r"""Representation of the operator as a product of other operators (static method).
@@ -532,16 +563,14 @@ class MultiControlledX(ControlledOp):
             wires (Iterable[Any] or Wires): wires that the operation acts on
             work_wires (Wires): optional work wires used to decompose
                 the operation into a series of Toffoli gates.
-            control_values (Union[bool, list[bool], int, list[int]]): The value(s) the control wire(s)
-                should take. Integers other than 0 or 1 will be treated as ``int(bool(x))``.
-
+            control_values (str): a string of bits representing the state of the control
+                wires to control on (default is the all 1s state)
         Returns:
             list[Operator]: decomposition into lower level operations
 
         **Example:**
 
-        >>> print(qml.MultiControlledX.compute_decomposition(
-        ...     wires=[0,1,2,3], control_values=[1,1,1], work_wires=qml.wires.Wires("aux")))
+        >>> print(qml.MultiControlledX.compute_decomposition(wires=[0,1,2,3],control_values="111", work_wires=qml.wires.Wires("aux")))
         [Toffoli(wires=[2, 'aux', 3]),
         Toffoli(wires=[0, 1, 'aux']),
         Toffoli(wires=[2, 'aux', 3]),
@@ -549,36 +578,121 @@ class MultiControlledX(ControlledOp):
 
         """
 
-        if len(wires) < 2:
-            raise ValueError(f"Wrong number of wires. {len(wires)} given. Need at least 2.")
-
-        target_wire = wires[-1]
-        control_wires = wires[:-1]
+        target_wire = wires[~0]
+        control_wires = wires[:~0]
 
         if control_values is None:
-            control_values = [True] * len(control_wires)
+            control_values = "1" * len(control_wires)
 
-        work_wires = work_wires or []
         if len(control_wires) > 2 and len(work_wires) == 0:
             raise ValueError(
                 "At least one work wire is required to decompose operation: MultiControlledX"
             )
 
-        flips1 = [qml.PauliX(wires=w) for w, val in zip(control_wires, control_values) if not val]
+        flips1 = [
+            qml.PauliX(control_wires[i]) for i, val in enumerate(control_values) if val == "0"
+        ]
 
         if len(control_wires) == 1:
-            decomp = [qml.CNOT(wires=wires)]
+            decomp = [qml.CNOT(wires=[control_wires[0], target_wire])]
         elif len(control_wires) == 2:
-            decomp = qml.Toffoli.compute_decomposition(wires=wires)
+            decomp = [qml.Toffoli(wires=[*control_wires, target_wire])]
         else:
-            decomp = decompose_mcx(control_wires, target_wire, work_wires)
+            num_work_wires_needed = len(control_wires) - 2
+
+            if len(work_wires) >= num_work_wires_needed:
+                decomp = MultiControlledX._decomposition_with_many_workers(
+                    control_wires, target_wire, work_wires
+                )
+            else:
+                work_wire = work_wires[0]
+                decomp = MultiControlledX._decomposition_with_one_worker(
+                    control_wires, target_wire, work_wire
+                )
 
-        flips2 = [qml.PauliX(wires=w) for w, val in zip(control_wires, control_values) if not val]
+        flips2 = [
+            qml.PauliX(control_wires[i]) for i, val in enumerate(control_values) if val == "0"
+        ]
 
         return flips1 + decomp + flips2
 
-    def decomposition(self):
-        return self.compute_decomposition(self.active_wires, self.work_wires, self.control_values)
+    @staticmethod
+    def _decomposition_with_many_workers(control_wires, target_wire, work_wires):
+        """Decomposes the multi-controlled PauliX gate using the approach in Lemma 7.2 of
+        https://arxiv.org/abs/quant-ph/9503016, which requires a suitably large register of
+        work wires"""
+        num_work_wires_needed = len(control_wires) - 2
+        work_wires = work_wires[:num_work_wires_needed]
+
+        work_wires_reversed = list(reversed(work_wires))
+        control_wires_reversed = list(reversed(control_wires))
+
+        gates = []
+
+        for i in range(len(work_wires)):
+            ctrl1 = control_wires_reversed[i]
+            ctrl2 = work_wires_reversed[i]
+            t = target_wire if i == 0 else work_wires_reversed[i - 1]
+            gates.append(qml.Toffoli(wires=[ctrl1, ctrl2, t]))
+
+        gates.append(qml.Toffoli(wires=[*control_wires[:2], work_wires[0]]))
+
+        for i in reversed(range(len(work_wires))):
+            ctrl1 = control_wires_reversed[i]
+            ctrl2 = work_wires_reversed[i]
+            t = target_wire if i == 0 else work_wires_reversed[i - 1]
+            gates.append(qml.Toffoli(wires=[ctrl1, ctrl2, t]))
+
+        for i in range(len(work_wires) - 1):
+            ctrl1 = control_wires_reversed[i + 1]
+            ctrl2 = work_wires_reversed[i + 1]
+            t = work_wires_reversed[i]
+            gates.append(qml.Toffoli(wires=[ctrl1, ctrl2, t]))
+
+        gates.append(qml.Toffoli(wires=[*control_wires[:2], work_wires[0]]))
+
+        for i in reversed(range(len(work_wires) - 1)):
+            ctrl1 = control_wires_reversed[i + 1]
+            ctrl2 = work_wires_reversed[i + 1]
+            t = work_wires_reversed[i]
+            gates.append(qml.Toffoli(wires=[ctrl1, ctrl2, t]))
+
+        return gates
+
+    @staticmethod
+    def _decomposition_with_one_worker(control_wires, target_wire, work_wire):
+        """Decomposes the multi-controlled PauliX gate using the approach in Lemma 7.3 of
+        https://arxiv.org/abs/quant-ph/9503016, which requires a single work wire"""
+        tot_wires = len(control_wires) + 2
+        partition = int(np.ceil(tot_wires / 2))
+
+        first_part = control_wires[:partition]
+        second_part = control_wires[partition:]
+
+        gates = [
+            MultiControlledX(
+                wires=first_part + work_wire,
+                work_wires=second_part + target_wire,
+            ),
+            MultiControlledX(
+                wires=second_part + work_wire + target_wire,
+                work_wires=first_part,
+            ),
+            MultiControlledX(
+                wires=first_part + work_wire,
+                work_wires=second_part + target_wire,
+            ),
+            MultiControlledX(
+                wires=second_part + work_wire + target_wire,
+                work_wires=first_part,
+            ),
+        ]
+
+        return gates
+
+    @property
+    def is_hermitian(self):
+        return True
```
</details>

File: pennylane/ops/op_math/controlled_decompositions.py
Comment (astralcai) [reply pennylane/ops/op_math/controlled_decompositions.py:1]: Basically changing all calls to the `Controlled` constructor to use `ctrl`, as well as the `MultiControlledX` constructor to use `ctrl` as well, because in some cases, if it is a `MultiControlledX` with only two control wires, both controlled on 1, it would become a `Toffoli` instead. Other than this I didn't change anything.

File: pennylane/ops/op_math/controlled_ops.py [hunk 9]
```diff
@@ -141,8 +146,8 @@ def __init__(
         self._name = "ControlledQubitUnitary"
 
     def _controlled(self, wire):
-        ctrl_wires = self.control_wires + wire
-        values = None if self.control_values is None else self.control_values + [True]
+        ctrl_wires = wire + self.control_wires
+        values = None if self.control_values is None else [True] + self.control_values
```
[the remark this answers is not in the document]
Comment (astralcai) [reply pennylane/ops/op_math/controlled_ops.py:150]: It wasn't, but now that it inherits from `Controlled`, this conflicts with the convention with all other controlled operations, which caused some test cases to fail, which was how I caught it.

[the remark this answers is not in the document]
Comment (astralcai) [reply pennylane/ops/op_math/controlled_ops.py:150]: I don't think it was, but after the clean-up, some tests were failing because of this. It would be consistent with how all the other nested controlled operations order their control wires.

File: pennylane/ops/op_math/controlled.py [hunk 10]
```diff
@@ -190,6 +188,111 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
+@functools.lru_cache()
+def _get_special_ops():
```
[the remark this answers is not in the document]
Comment (astralcai) [reply pennylane/ops/op_math/controlled.py:192]: I believe it was circular import errors

File: tests/ops/op_math/test_controlled.py (hunk 4 above)
Comment (albi3ro) [reply tests/ops/op_math/test_controlled.py:2024]: So `qml.simplify` and `Adjoint.decomposition` would still turn `Adjoint(Controlled)` into `Controlled(Adjoint)` simply by updating the adjoint method.  So we would still be able to handle `Adjoint(Controlled)` just fine. The idea is "when" do we handle this? Lazily, or when requested in a decomposition?

Comment (astralcai) [reply tests/ops/op_math/test_controlled.py:2024]: I think similar to what I did in this PR, I would hope that `Adjoint(Controlled)` and `Controlled(Adjoint)` would decompose into the same thing, instead of having `Adjoint(Controlled)` decompose into `Controlled(Adjoint)`, unless I misunderstood what you mean.

File: pennylane/ops/op_math/controlled.py [hunk 11]
```diff
@@ -190,6 +196,115 @@ def wrapper(*args, **kwargs):
...
+
+    if isinstance(op, qml.QubitUnitary):
+        return qml.ControlledQubitUnitary(
+            op, control_wires=control, control_values=control_values, work_wires=work_wires
+        )
+
+    # A controlled ControlledPhaseShift should not be compressed to a multi controlled
+    # PhaseShift because the decomposition of PhaseShift contains a GlobalPhase that we
+    # do not have a controlled version of.
+    # TODO: remove this special case when we support ControlledGlobalPhase
```
Comment (timmysilv) [inline pennylane/ops/op_math/controlled.py:251]:
```suggestion
    # TODO: remove this special case when we support ControlledGlobalPhase (sc-44933)
```

**Review by timmysilv — APPROVED** [review]
awesome!
